File: generate_data.py
import pandas as pd
import numpy as np
import os

# Create required directories
os.makedirs('data', exist_ok=True)

# Set random seed for reproducibility
np.random.seed(42)

# Goal: Generate approximately 10,000 rows of data.
N = 10000

# 1. Generate random dates between 2015-01-01 and 2025-12-31
start_date = np.datetime64('2015-01-01')
end_date = np.datetime64('2025-12-31')
date_range = (end_date - start_date).astype(int)
random_days = np.random.randint(0, date_range, size=N)
dates = start_date + random_days

# Extract Year for Streamlit dashboard compatibility
years = pd.to_datetime(dates).year

# 2. Assign Regions
regions = np.random.choice(
    ['North America', 'Europe', 'China', 'Rest of World'], 
    size=N, 
    p=[0.25, 0.3, 0.4, 0.05]
)

# 3. Assign Manufacturers
manufacturers = np.random.choice(
    ['Tesla', 'BYD', 'Volkswagen', 'Hyundai', 'General Motors', 'Ford', 'NIO', 'Rivian', 'Others'], 
    size=N
)

# 4. Assign Powertrain
powertrains = np.random.choice(['BEV', 'PHEV'], size=N, p=[0.75, 0.25])

# 5. Assign Sales volume per record (batch size)
# To keep the realistic exponential growth trend over the years:
growth_factor = 1.35 ** (years - 2015)
base_sales = np.random.randint(5, 50, size=N)

# Apply manufacturer multipliers based on region and year
maker_bias = np.ones(N)
for i in range(N):
    y = years[i]
    r = regions[i]
    m = manufacturers[i]
    b = 1.0
    
    if m == 'Tesla' and r in ['North America', 'Europe']: b *= 2.5
    elif m == 'BYD' and r == 'China': b *= 3.0
    elif m in ['Volkswagen', 'General Motors', 'Ford'] and r in ['Europe', 'North America']: b *= 1.8
    elif m == 'NIO' and r == 'China': b *= 1.5
    elif m == 'Rivian' and r == 'North America' and y >= 2021: b *= 1.5
    elif m in ['Rivian', 'NIO'] and y < 2018: b *= 0.1
    
    maker_bias[i] = b

final_sales = (base_sales * growth_factor * maker_bias).astype(int)
# Ensure at least 1 sale
final_sales = np.where(final_sales < 1, 1, final_sales)

df = pd.DataFrame({
    'Date': dates,
    'Year': years,
    'Region': regions,
    'Manufacturer': manufacturers,
    'Powertrain': powertrains,
    'Sales': final_sales
})

# Sort chronologically
df = df.sort_values('Date')

# Export to CSV
df.to_csv('data/ev_sales_data.csv', index=False)
print(f"Dataset successfully updated with {len(df)} rows at data/ev_sales_data.csv")

File: generate_data.js
const fs = require('fs');
const path = require('path');

// Ensure data directory exists
const dir = path.join(__dirname, 'data');
if (!fs.existsSync(dir)){
    fs.mkdirSync(dir);
}

const N = 10000;
const startDate = new Date('2015-01-01');
const endDate = new Date('2025-12-31');
const dateRange = endDate.getTime() - startDate.getTime();

const regionsDist = [
    { name: 'North America', prob: 0.25 },
    { name: 'Europe', prob: 0.30 },
    { name: 'China', prob: 0.40 },
    { name: 'Rest of World', prob: 0.05 }
];

const manufacturers = ['Tesla', 'BYD', 'Volkswagen', 'Hyundai', 'General Motors', 'Ford', 'NIO', 'Rivian', 'Others'];

function getRandomRegion() {
    let r = Math.random();
    let sum = 0;
    for (let i = 0; i < regionsDist.length; i++) {
        sum += regionsDist[i].prob;
        if (r <= sum) return regionsDist[i].name;
    }
    return 'Rest of World';
}

function getRandomManufacturer() {
    return manufacturers[Math.floor(Math.random() * manufacturers.length)];
}

const stream = fs.createWriteStream(path.join(dir, 'ev_sales_data.csv'));
stream.write('Date,Year,Region,Manufacturer,Powertrain,Sales\n');

for (let i = 0; i < N; i++) {
    const randomTime = startDate.getTime() + Math.random() * dateRange;
    const date = new Date(randomTime);
    const year = date.getFullYear();
    const formattedDate = date.toISOString().split('T')[0];
    
    const region = getRandomRegion();
    const manufacturer = getRandomManufacturer();
    const powertrain = Math.random() < 0.75 ? 'BEV' : 'PHEV';
    
    const growthFactor = Math.pow(1.35, year - 2015);
    let baseSales = Math.floor(Math.random() * (50 - 5 + 1)) + 5;
    
    let b = 1.0;
    if (manufacturer === 'Tesla' && ['North America', 'Europe'].includes(region)) b *= 2.5;
    else if (manufacturer === 'BYD' && region === 'China') b *= 3.0;
    else if (['Volkswagen', 'General Motors', 'Ford'].includes(manufacturer) && ['Europe', 'North America'].includes(region)) b *= 1.8;
    else if (manufacturer === 'NIO' && region === 'China') b *= 1.5;
    else if (manufacturer === 'Rivian' && region === 'North America' && year >= 2021) b *= 1.5;
    else if (['Rivian', 'NIO'].includes(manufacturer) && year < 2018) b *= 0.1;
    
    let finalSales = Math.floor(baseSales * growthFactor * b);
    if (finalSales < 1) finalSales = 1;
    
    stream.write(`${formattedDate},${year},${region},${manufacturer},${powertrain},${finalSales}\n`);
}

stream.end();
console.log(`Dataset successfully updated with ${N} rows at data/ev_sales_data.csv`);
